### DISCUSSION TITLE
[@types/node] Node v16 timers.d.ts (setTimeout, setImmediate etc.) no longer works for certain callbacks (e.g. Promise callbacks without a specified type)

### DISCUSSION THREAD

**Discussion by birtles** [DefinitelyTyped/DefinitelyTyped#54258] [Issues with a @types package]
I just updated to @types/node 16.0.0 and I get errors from code that used to work like:

```ts
setImmediate(res);
// ^Expected 2 arguments, but got 1.ts(2554)
// timers.d.ts(54, 88): Arguments for the rest parameter 'args' were not provided.
```

It looks like 15.12.5 defined `setImmediate` like this:

```ts
declare function setImmediate(callback: (...args: any[]) => void, ...args: any[]): NodeJS.Immediate;
```

but 16.0.0 defines it like this:

```ts
function setImmediate<TArgs extends any[]>(callback: (...args: TArgs) => void, ...args: TArgs): NodeJS.Immediate;
```

Note that `TArgs` does not have any default value so it is resolved to `unknown` in the code sample above.

I can work around using:

```ts
setImmediate<any[]>(res);
```

but that seems a bit clumsy.

TypeScript version: 4.3.5

**Comment by Zielak** [comment]
`setTimeout` suffers with similar issue :(

```
setTimeout(func, time)
---
Expected 3 arguments, but got 2.ts(2554)
timers.d.ts(42, 99): Arguments for the rest parameter 'args' were not provided.
function setTimeout<[value: unknown]>(callback: (value: unknown) => void, ms?: number, value: unknown): NodeJS.Timeout
namespace setTimeout
```

**Comment by birtles** [comment]
I tried to write a PR for this but it's more complicated than I thought. Here is the additional test case that fails:

```ts
// unresolved callback argument types
{
    new Promise(resolve => {
        timers.setImmediate(resolve);
    });
}
```

(I couldn't write a failing test using `setTimeout` because the project ends up picking up the definition for `setTimeout` in `lib.dom.d.ts` and resolving it using that. I'm not sure how to get TS to ignore that definition. Even if I try using `timers.setTimeout` or `global.setTimeout` it still picks it up.)

In this case, `resolve` has the type `(value: unknown) => void` so `TArgs` is resolved as `[unknown]` and hence TS expects a single `args` parameter of type `unknown` and complains when it's not found.

This can be fixed by setting the type of `Promise`:

```ts
    new Promise<void>(resolve => {
        timers.setImmediate(resolve);
    });
```

That said, I know a lot of people use Promises in this way together with `setTimeout` and `setImmediate` so it would be good if this continued to work without having to set the type of `Promise`.

Unfortunately I don't have enough TS knowledge to know how to go about doing that.

**Comment by birtles** [comment]
I ended up just adding an overload for the Promise case in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54286

I'm really not sure if that's the best approach or not, however.
